src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java

```diff
@@ -46,14 +59,103 @@ private boolean checkIfProductsCacheNeedsUpdate() {
     }
 
     @Override
-    public ProductList findBySearchRequest(String searchRequest) {
-        var result = getProductListCache().getProducts()
-                .stream()
+    public ProductList findBySearchRequest(SearchModel searchModel) {
+        List<Product> result = executeSearch(searchModel.getSearchRequest());
+        return getSearchResultsOrAllProducts(result);
+    }
+
+    @Override
+    public ProductList searchForProductDescription(List<Product> products, SearchModel searchModel) {
+        setAppliedFiltersToSearchModel(searchModel);
+        List<Product> result = executeDescriptionSearch(products, searchModel.getDescription());
+        return getSearchResultsOrAllProducts(result);
+    }
+
+    @Override
+    public void setPriceFilters(List<Product> products) {
+        searchService.getSearchModel().setSortAscendingPrice(true);
+        List<Product> sortedProducts = sortListByPrice(products, searchService.getSearchModel()).getProducts();
+        BigDecimal priceLow = new BigDecimal("0");
+        BigDecimal priceHigh = sortedProducts.get(sortedProducts.size() - 1).getPrice();
+        searchService.setPriceLow(priceLow);
+        searchService.setPriceHigh(priceHigh);
+    }
+
+    @Override
+    public ProductList filterByPrice(List<Product> products, SearchModel searchModel) {
+        setAppliedFiltersToSearchModel(searchModel);
+        List<Product> result = products
+                .parallelStream()
+                .filter(applyPriceFilter(searchModel))
+                .collect(Collectors.toUnmodifiableList());
+        return getSearchResultsOrAllProducts(result);
+    }
+
+    private ProductList getSearchResultsOrAllProducts(List<Product> result) {
+        searchService.setSearchFailure(result.isEmpty());
+        return result.isEmpty()
+                ? new ProductList(getProductListCache().getProducts())
+                : new ProductList(result);
+    }
+
+    private void setAppliedFiltersToSearchModel(SearchModel searchModel) {
+        searchModel.setAppliedFiltersHeader(" with the applied filters");
+        String searchRequest = (null != searchService.getSearchModel().getSearchRequest())
```

Comment:
boolean hasSearchRequest() is a more readable name for this if statement.

Reply:
extracted